# narst/stylo: Output

## Introduction

This document describes the output produced by the pipeline. 

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

<!-- TODO nf-core: Write this documentation describing your workflow's output -->

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes data using the following steps:

- [nanoq](#nanoq) - Read filtering
- [rasusa](#rasusa) - Read Downsampling
- [Flye](#flye) - Assembly
- [Circlator](#circlator) - Circularization of assembly
- [Medaka](#medaka) - Assembly error correction
<!-- TODO: SOCRU -->
- [busco](#busco) - Assembly QC
- [Pipeline information](#pipeline-information) - Report metrics generated during the workflow execution

### nanoq

<details markdown="1">
<summary>Output files</summary>

- `nanoq/`
  - `*_filtered.fastq.gz`: Filtered reads
  - `*_filtered.stats`: basic stats summary for filtered reads

</details>

For further reading and documentation see [nanoq usage](https://github.com/esteinig/nanoq?tab=readme-ov-file#usage).

### rasusa

<details markdown="1">
<summary>Output files</summary>

- `rasusa/`
  - `*.fastq.gz`: downsampled reads

</details>

For further reading and documentation see [rasusa usage](https://github.com/mbhall88/rasusa?tab=readme-ov-file#usage).

### Flye

<details markdown="1">
<summary>Output files</summary>

- `flye/`
  - `*.assembly.fasta.gz` : Final assembly
  - `*.assembly_graph.gfa.gz` : Final repeat graph in gfa format
  - `*.assembly_graph.gv.gz` : Final repeat graph in gv format
  - `*.assembly_info.txt` : basic info about assembly
  - `*.flye.log` : full log for flye run
  - `*.params.json`: input parameters used for flye

</details>

For further reading and documentation see the [Flye usage page](https://github.com/mikolmogorov/Flye/blob/flye/docs/USAGE.md).

### Circlator

<details markdown="1">
<summary>Output files</summary>

- `circlator_fixstart/`
  - `*.fasta`: circularized assembly

</details>

For further reading and documentation see [circlator usage](https://github.com/sanger-pathogens/circlator?tab=readme-ov-file#usage).

### Medaka

<details markdown="1">
<summary>Output files</summary>

- `medaka/`
  - `*.fa.gz`: Error corrected assembly

</details>

For further reading and documentation see [medaka usage](https://github.com/nanoporetech/medaka?tab=readme-ov-file#usage).

<!-- TODO: SOCRU -->



### busco

<details markdown="1">
<summary>Output files</summary>

- `busco/`
  - `*-auto-busco/`: Directory containing all intermediate outputs of busco
  - `*-auto-busco.batch_summary.txt`: full QC summary
  - `short_summary.generic.*` : QC summary for generic organisms (bacteria, ...)
  - `short_summary.specific.*` : QC summary for specific organisms (enterobacteriaceae, campylobacterota, ...)

</details>

For further reading and documentation see the [busco github](https://github.com/metashot/busco).

<!-- EXAMPLE
### FastQC

<details markdown="1">
<summary>Output files</summary>

- `fastqc/`
  - `*_fastqc.html`: FastQC report containing quality metrics.
  - `*_fastqc.zip`: Zip archive containing the FastQC report, tab-delimited data file and plot images.

</details>

[FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) gives general quality metrics about your sequenced reads. It provides information about the quality score distribution across your reads, per base sequence content (%A/T/G/C), adapter contamination and overrepresented sequences. For further reading and documentation see the [FastQC help pages](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/).

![MultiQC - FastQC sequence counts plot](images/mqc_fastqc_counts.png)

![MultiQC - FastQC mean quality scores plot](images/mqc_fastqc_quality.png)

![MultiQC - FastQC adapter content plot](images/mqc_fastqc_adapter.png)

:::note
The FastQC plots displayed in the MultiQC report shows _untrimmed_ reads. They may contain adapter sequence and potentially regions with low quality.
:::

### MultiQC

<details markdown="1">
<summary>Output files</summary>

- `multiqc/`
  - `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
  - `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
  - `multiqc_plots/`: directory containing static images from the report in various formats.

</details>

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarising all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

Results generated by MultiQC collate pipeline QC from supported tools e.g. FastQC. The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability. For more information about how to use MultiQC reports, see <http://multiqc.info>.

 -->

### Pipeline information

<details markdown="1">
<summary>Output files</summary>

- `pipeline_info/`
  - Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  - Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameter's are used when running the pipeline.
  - Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.
  - Parameters used by the pipeline run: `params.json`.

</details>

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.
