# ncezid-biome/stylo: Output

## Introduction

This document describes the output produced by the pipeline. 

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes data using the following steps:

- [nanoq](#nanoq) - Read filtering
- [Rasusa](#rasusa) - Read Downsampling
- [Flye](#flye) - Assembly
- [Dnaapler](#dnaapler) - Reorientation of contigs
- [Medaka](#medaka) - Assembly error correction
- [QUAST](#quast) - Assembly QC
- [Pipeline information](#pipeline-information) - Report metrics generated during the workflow execution

### nanoq

<details markdown="1">
<summary>Output files</summary>

- `nanoq/`
  - `*_filtered.fastq.gz`: filtered reads
  - `*_filtered.stats`: basic stats summary for filtered reads

</details>

Ultra-fast quality control and summary reports for nanopore reads. For further reading and documentation see [nanoq usage](https://github.com/esteinig/nanoq?tab=readme-ov-file#usage).

### Rasusa

<details markdown="1">
<summary>Output files</summary>

- `rasusa/`
  - `*.fastq.gz`: downsampled reads

</details>

Randomly subsample sequencing reads to a specified coverage. For further reading and documentation see [rasusa usage](https://github.com/mbhall88/rasusa?tab=readme-ov-file#usage).

### Flye

<details markdown="1">
<summary>Output files</summary>

- `flye/`
  - `*.assembly.fasta.gz` : final assembly
  - `*.assembly_graph.gfa.gz` : final repeat graph in gfa format
  - `*.assembly_graph.gv.gz` : final repeat graph in gv format
  - `*.assembly_info.txt` : basic info about assembly
  - `*.flye.log` : full log for flye run
  - `*.params.json`: input parameters used for flye

</details>

Fast and accurate de novo assembler for single molecule sequencing reads. For further reading and documentation see the [Flye usage page](https://github.com/mikolmogorov/Flye/blob/flye/docs/USAGE.md).

### Dnaapler

<details markdown="1">
<summary>Output files</summary>

- `dnaapler/`
  - `*reoriented.fasta`: reoriented assembly
  - `*MMseqs2_output.txt`: MMseqs output
  - `*all_reorientation_summary.tsv`: reorientation summery
  - `*.log`: log

</details>

For further reading and documentation see [dnaapler usage](https://github.com/gbouras13/dnaapler?tab=readme-ov-file#usage).

### Medaka

<details markdown="1">
<summary>Output files</summary>

- `medaka/`
  - `*.fa.gz`: Error corrected assembly

</details>

Neural network sequence error correction. For further reading and documentation see [medaka usage](https://github.com/nanoporetech/medaka?tab=readme-ov-file#usage).

### QUAST

<details markdown="1">
<summary>Output files</summary>

- `quast/`
  - `*.tsv`,`*/report.tsv`,`*/transposed_report.tsv`: Final summary of assembly stats in tsv format
  - `*/report.tex`,`*/transposed_report.tex`: Final summary of assembly stats in TeX format
  - `*/report.txt`,`*/transposed_report.txt`: Final summary of assembly stats in an easy to read txt format

</details>

QUAST provides evaluates genome assemblies by computing various metrics, such as N50, L50, GC content, length and contig distributions ... 

### Pipeline information

<details markdown="1">
<summary>Output files</summary>

- `pipeline_info/`
  - Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  - Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameter's are used when running the pipeline.
  - Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.
  - Parameters used by the pipeline run: `params.json`.

</details>

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.
